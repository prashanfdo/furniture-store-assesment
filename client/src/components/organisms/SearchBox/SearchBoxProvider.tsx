import constate from "constate";
import useSearchSuggestionsQuery from "queries/useSearchSuggestionsQuery";
import { useState } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";

// TODO: add comments
const [SearchBoxProvider, useSearchBoxContext] = constate(() => {
  const navigate = useNavigate();
  const [showSuggestions, setShowSuggestions] = useState<boolean>(false);

  const [searchParams] = useSearchParams();
  const initialSearchText = searchParams.get("q") || "";
  const [searchText, setSearchText] = useState<string>("");

  const handleSearchSubmit = (searchText: string) => {
    setShowSuggestions(false);
    if (!!searchText?.trim()) {
      navigate(`/search?q=${encodeURIComponent(searchText?.trim())}`);
    } else {
      navigate(`/`);
    }
  };

  const handleSearchChange = (newText: string) => {
    if (newText !== searchText) {
      setShowSuggestions(true);
    }
    setSearchText(newText);
  };

  // TODO: handle loading and error states
  const { data } = useSearchSuggestionsQuery(searchText);

  const searchSuggestions = data?.searchSuggestions || [];

  return {
    initialSearchText,
    handleSearchChange,
    handleSearchSubmit,
    searchSuggestions,
    showSuggestions,
    setShowSuggestions,
  };
});

export { useSearchBoxContext };
export default SearchBoxProvider;
