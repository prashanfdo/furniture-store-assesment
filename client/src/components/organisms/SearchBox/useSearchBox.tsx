import useSearchSuggestionsQuery from "queries/useSearchSuggestionsQuery";
import { useState } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";

// TODO: add comments
const useSearchBox = () => {
  const navigate = useNavigate();
  const [showSuggestions, setShowSuggestions] = useState<boolean>(false);

  const [searchParams] = useSearchParams();
  const initialSearchText = searchParams.get("q") || "";
  const [searchText, setSearchText] = useState<string>("");

  const handleSearchSubmit = (searchText: string) => {
    setShowSuggestions(false);
    if (!!searchText?.trim()) {
      navigate(`/search?q=${encodeURIComponent(searchText?.trim())}`);
    } else {
      navigate(`/`);
    }
  };

  const handleSearchChange = (newText: string) => {
    if (newText !== searchText) {
      setShowSuggestions(true);
    }
    setSearchText(newText);
  };

  // TODO: handle loading and error states
  const { data, loading, error } = useSearchSuggestionsQuery(searchText);

  const searchSuggestions = data?.searchSuggestions || [];

  return {
    loading,
    error,
    initialSearchText,
    handleSearchChange,
    handleSearchSubmit,
    searchSuggestions,
    showSuggestions,
    setShowSuggestions,
  };
};

export default useSearchBox;
