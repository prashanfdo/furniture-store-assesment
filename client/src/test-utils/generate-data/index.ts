export * from "./article";
export * from "./category";
export * from "./search";
export * from "./gql";
