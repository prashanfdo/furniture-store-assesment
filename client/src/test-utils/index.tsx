import { render, RenderOptions, renderHook, RenderHookOptions, RenderHookResult } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import user from "@testing-library/user-event";
import { ProvidersForTests } from "providers";

type GraphQlMocks = React.ComponentProps<typeof ProvidersForTests>["mocks"];

type WrapperProps = React.ComponentProps<typeof ProvidersForTests> & {
  children: React.ReactNode;
};
const Wrapper: FC<WrapperProps> = ({ children, mocks, initialRoute }: WrapperProps) => {
  return (
    <ProvidersForTests mocks={mocks} initialRoute={initialRoute}>
      {children}
    </ProvidersForTests>
  );
};
const wrapper =
  (mocks: GraphQlMocks, initialRoute?: string) =>
  ({ children }: { children: React.ReactNode }) =>
    (
      <Wrapper mocks={mocks} initialRoute={initialRoute}>
        {children}
      </Wrapper>
    );

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & { gqlMocks: GraphQlMocks; initialRoute: string },
) => {
  const view = render(ui, { wrapper: wrapper(options?.gqlMocks, options?.initialRoute), ...options });
  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};

export const customRenderHook = <T, R>(
  callback: (props: T) => R,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { gqlMocks?: GraphQlMocks; initialRoute?: string },
): RenderHookResult<R, T> => {
  return renderHook(callback, { ...options, wrapper: wrapper(options?.gqlMocks, options?.initialRoute) });
};

export { faker } from "@faker-js/faker";
export * from "@testing-library/react";
export * from "./generate-data";
export { customRender as render };
export { customRenderHook as renderHook };
export { user };
