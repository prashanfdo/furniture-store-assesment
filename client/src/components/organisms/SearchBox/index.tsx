import { SearchBoxForm } from "components/molecules";
import React from "react";
import useSearchBox from "./useSearchBox";

const SearchBox: React.FC = () => {
  const { handleSearchChange, initialSearchText, handleSearchSubmit, searchSuggestions, showSuggestions } =
    useSearchBox();

  return (
    <SearchBoxForm
      showSuggestions={showSuggestions}
      initialSearchText={initialSearchText}
      onSearchSubmit={handleSearchSubmit}
      onSearchChange={handleSearchChange}
      suggestions={searchSuggestions}
    />
  );
};

export default SearchBox;
