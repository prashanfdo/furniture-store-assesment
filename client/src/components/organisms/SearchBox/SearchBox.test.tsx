import { render } from "test-utils";
import SearchBox from ".";
import useSearchBox from "./useSearchBox";

jest.mock("./useSearchBox");
jest.mock("components/molecules", () => {
  return {
    SearchBoxForm: () => <div data-testid="search-box-form" />,
  };
});
const mockuseSearchBox = useSearchBox as jest.MockedFunction<typeof useSearchBox>;

describe("SearchBox", () => {
  beforeEach(() => {
    mockuseSearchBox.mockReset();
  });

  it("should render", async () => {
    mockuseSearchBox.mockReturnValue({
      initialSearchText: "test search",
      onSearchChange: jest.fn(),
    } as never);
    const { container } = render(<SearchBox />);
    expect(container).toMatchSnapshot();
  });
});

export default useSearchBox;
