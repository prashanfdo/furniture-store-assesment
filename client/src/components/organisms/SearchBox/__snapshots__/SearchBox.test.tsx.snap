// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox should render 1`] = `
<div>
  <div
    data-testid="search-box-form"
  />
</div>
`;
